### PR TITLE
Check void return types on closures

### PR DESCRIPTION
**Pull request by pepakriz**

### CODE REVIEW

File: SlevomatCodingStandard/Sniffs/TypeHints/TypeHintDeclarationSniff.php [hunk 1]
```diff
@@ -473,6 +478,51 @@ private function checkReturnTypeHints(\PHP_CodeSniffer\Files\File $phpcsFile, in
...
+	private function checkClosure(\PHP_CodeSniffer\Files\File $phpcsFile, int $closurePointer)
+	{
+		$returnTypeHint = FunctionHelper::findReturnTypeHint($phpcsFile, $closurePointer);
+		$returnsValue = FunctionHelper::returnsValue($phpcsFile, $closurePointer);
+
+		if (!$returnsValue && $returnTypeHint !== null && $returnTypeHint->getTypeHint() !== 'void') {
+			$fix = $phpcsFile->addFixableError(
+				'Closure λ() has useless return type hint.',
+				$closurePointer,
+				self::CODE_USELESS_RETURN_TYPE_HINT
```
Comment (pepakriz) [inline SlevomatCodingStandard/Sniffs/TypeHints/TypeHintDeclarationSniff.php]: maybe better naming: `CODE_INCORRECT_RETURN_TYPE_HINT`

File: SlevomatCodingStandard/Sniffs/TypeHints/TypeHintDeclarationSniff.php [hunk 2]
```diff
@@ -473,6 +478,51 @@ private function checkReturnTypeHints(\PHP_CodeSniffer\Files\File $phpcsFile, in
...
 	}
 
+	private function checkClosure(\PHP_CodeSniffer\Files\File $phpcsFile, int $closurePointer)
+	{
+		$returnTypeHint = FunctionHelper::findReturnTypeHint($phpcsFile, $closurePointer);
+		$returnsValue = FunctionHelper::returnsValue($phpcsFile, $closurePointer);
+
+		if (!$returnsValue && $returnTypeHint !== null && $returnTypeHint->getTypeHint() !== 'void') {
+			$fix = $phpcsFile->addFixableError(
+				'Closure λ() has useless return type hint.',
```
Comment (Majkl578) [inline SlevomatCodingStandard/Sniffs/TypeHints/TypeHintDeclarationSniff.php]: This message is confusing, there is no `λ()` function/closure.

File: SlevomatCodingStandard/Sniffs/TypeHints/TypeHintDeclarationSniff.php [hunk 3]
```diff
@@ -473,6 +478,51 @@ private function checkReturnTypeHints(\PHP_CodeSniffer\Files\File $phpcsFile, in
...
+				$phpcsFile->fixer->replaceToken($closeParenthesisPosition, $this->enableVoidTypeHint ? '): void ' : ') ');
+				$phpcsFile->fixer->endChangeset();
+			}
+
+			return;
+		}
+
+		if ($this->enableVoidTypeHint && !$returnsValue && $returnTypeHint === null) {
+			$fix = $phpcsFile->addFixableError(
+				'Closure λ() does not have void return type hint.',
```
Comment (Majkl578) [inline SlevomatCodingStandard/Sniffs/TypeHints/TypeHintDeclarationSniff.php]: dtto

**Comment by pepakriz** [comment]
Sorry for being late. Now it is ready for review.

**Comment by kukulich** [comment]
Thanks. Cherrypicked because of wrong error message: https://github.com/slevomat/coding-standard/pull/171/files#diff-7e99248e85bad968f95d2da8ede2b859R488

**Pull request closed without merging**